stumpy/scrump.py

```diff
@@ -645,48 +728,56 @@ def update(self):
                 self._T_B_subseq_isconstant,
                 self._diags[start_idx:stop_idx],
                 self._ignore_trivial,
+                self._k,
             )
 
-            # Update matrix profile and indices
-            for i in range(self._P.shape[0]):
-                if self._P[i, 0] > P[i, 0]:
-                    self._P[i, 0] = P[i, 0]
-                    self._I[i, 0] = I[i, 0]
-                # left matrix profile and left matrix profile indices
-                if self._P[i, 1] > P[i, 1]:
-                    self._P[i, 1] = P[i, 1]
-                    self._I[i, 1] = I[i, 1]
-                # right matrix profile and right matrix profile indices
-                if self._P[i, 2] > P[i, 2]:
-                    self._P[i, 2] = P[i, 2]
-                    self._I[i, 2] = I[i, 2]
+            # Update (top-k) matrix profile and indices
+            core._merge_topk_PI(self._P, P, self._I, I)
+
+            # update left matrix profile and indices
+            cond = PL < self._PL
+            self._PL = np.where(cond, PL, self._PL)
+            self._IL = np.where(cond, IL, self._IL)
+
+            # update right matrix profile and indices
+            cond = PR < self._PR
+            self._PR = np.where(cond, PR, self._PR)
+            self._IR = np.where(cond, IR, self._IR)
 
             self._chunk_idx += 1
 
     @property
     def P_(self):
         """
-        Get the updated matrix profile
+        Get the updated (top-k) matrix profile. When `k=1`, it is a 1d array.
+        When `k>1`, it is a 2d array with exactly k columns consist of (top-k) matrix
+        profile.
         """
-        return self._P[:, 0].astype(np.float64)
+        if self._k == 1:
```

Comment:
This is to make sure the structure of `self._P` does not change for k=1. (same goes for `self._I`) What do you think?

Reply:
Hmmm, I understand that this will be a breaking change but, frankly, I don't think many people (if any??) are using `scrump` and so I doubt that we need this. I say we remove it but we need to make sure that we are consistent and make ALL matrix profile outputs 2D arrays. Can you make a note of this? Maybe even a new issue. Additionally, this breaking change means that we'll need to double check all of our tutorials after as well as they would break too.

Reply:
> I don't think many people (if any??) are using `scrump` and so I doubt that we need this. I say we remove it 

Sure. We can keep it as 2D array. 

> but we need to make sure that we are consistent and make ALL matrix profile outputs 2D arrays. Can you make a note of this? Maybe even a new issue. Additionally, this breaking change means that we'll need to double check all of our tutorials after as well as they would break too.

I think `scrump` is the only place that we have to deal with this issue. In other parts, we return `mp`, which has shape `(l, 4)` when `k=1`. So, are you thinking of breaking those structures and returns two outputs `P` and `I` in `stump`/`stumped`/`gpu_stump`? 

Or, 
since there are probably a few people that use `scrump` as of now, we may just change the output structure for `scrump`?
```
self.mp_ = np.column_stack((P, I, IL, IR))
```
Then, there would no problem and we just need to change the tutorial(s) related to `scrump`.

Reply:
> So, are you thinking of breaking those structures and returns two outputs P and I in stump/stumped/gpu_stump?

No, I think it is fine for `scrump` to have `P_` and `I_` attributes and, separately, I think it's fine that `stump` (and friends) return matrix profiles along with the indices as one 2D array. In my mind, the difference is that `scrump` is a class object and can therefore have class attributes and maintains state. However, `stump` (and friends) are functions and so should remain simple with less sophisticated return values.

> I think scrump is the only place that we have to deal with this issue. 

Actually, I was thinking about `stumpi` and whether we need to add top-k there too?

Reply:
> In my mind, the difference is that `scrump` is a class object and can therefore have class attributes and maintains state. However, `stump` (and friends) are functions and so should remain simple with less sophisticated return values.

I see. Thanks for the explanation.


> Actually, I was thinking about `stumpi` and whether we need to add top-k there too?

I did not know about `stumpi`! I just checked it out. I think we can add it for the sake of consistency. Not sure if users want to use it or if VALMOD method can be applied to it easily...but for now, there should be no problem (other than increasing the time of testing)

Reply:
Sure. I will modify it.